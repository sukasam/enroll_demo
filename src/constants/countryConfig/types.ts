import { Alpha2, Alpha3, Currency, optInNotificationTypes } from "./enums";

export type Language = {
    code: string;
    name: string;
};

export interface PackPerk {
    icon: string;
    perk: string;
}

export interface PackExtra {
    icon: string;
    title: string;
    description: string;
}

export interface PackProduct {
    image: string;
    title: string;
    description: string;
}

export interface VariantOption {
    // Id within option to distinguish between values. Each option has its own set of ids.
    interOptionId: number;
    title: string;
    description?: string;
    disclaimer?: string;
    limitedColor?: string;
    image?: string;
}

export interface Variant {
    sku: string;
    name: string;
    description: string;
    productImage: any;
    partsCount: string;
    tabImage: any;
    options: {
        [key: string]: VariantOption;
    };
    price: number;
    rrp: number;
}

export interface DigitalPack {
    name: string;
    isDigital: true;
    buttonName: string;
    description: string;
    showPaymentType: boolean;
}

export interface CompleteDigitalPack extends DigitalPack {
    variants: Variant[];
}

export interface Pack {
    mainFeatures: string[];
    accentColor: string;
    buttonName: string;
    kitDescription: string;
    backgroundColor: string;
    perks: PackPerk[];
    extras: PackExtra[];
    products?: PackProduct[];
    annotations: string;
    showPaymentType: boolean;
    isDigital: false;
}

export interface CompletePack extends Pack {
    variants: Variant[];
}

export type ProductConfig = {
    recommendedPack: number | null;
    packs: (Pack | DigitalPack)[];
};

export type CompleteProductConfig = {
    recommendedPack: number;
    packs: (CompletePack | CompleteDigitalPack)[];
};

export interface WillCallAddress {
    address1?: string;
    address2?: string;
    city?: string;
    state?: string;
    zip?: string;
    country?: string;
    mapLink?: string;
    mapLinkKey?: string;
    rosettaKey?: string;
}

export type PaymentOptions = {
    hasPayPal?: boolean;
    hasCash?: boolean;
    hasBankWire?: boolean;
    hasCreditCard?: boolean;
    hasSavedCreditCard?: boolean;
    hasKlarna?: boolean;
};

export type MarketExceptions = {
    zipLength?: number | number[] | null;
    zipPattern?: RegExp | null;
    taxIdPattern?: RegExp;
    secondaryTaxIdPattern?: RegExp;
    taxInclusiveFrontEnd?: boolean;
    taxInclusiveBackEnd?: boolean;
    hasSecondaryTaxId?: boolean;
    secondaryTaxIdData?: {
        label: string;
    }[];
    disableTaxId?: boolean;
    hiddenShippingProvince?: boolean;
    useQuotesToValidateAddress?: boolean;
    hasCommissionPaymentOptions?: boolean;
    showOnboardingGuide?: boolean;
    showVisaAgreement?: boolean;
    hiddenShippingQuotesCity?: boolean;
    hiddenShippingQuotesZip?: boolean;
    willCallAddresses?: WillCallAddress[];
    thankYouImage?: string;
    hideUnicityPhone?: boolean;
    paymentOptions?: PaymentOptions;
    showStatesField?: boolean;
    useWorldPay?: boolean;
    directWorldPay?: boolean;
    usePayPal3ds?: boolean;
    notificationType?: optInNotificationTypes | undefined;
    onlyHasWillCall?: boolean;
    showOnlineRegistration?: boolean;
    hiddenShippingCity?: boolean;
    hiddenTooltip?: boolean;
    hiddenDecimal?: boolean;
    updateMainAddress?: boolean;
    showMemberApplicationGuide?: boolean;
    useGlobalFormatCurrency?: boolean;
    hiddenEnrollerId?: boolean;
    hiddenPersonalData?: boolean;
    useCookie?: boolean;
    formCreateAccount?: string[];
    formUpdateAccount?: string[];
};

type BaseCountryConfig = {
    alpha3: Alpha3;
    alpha2: Alpha2;
    name: string;
    languages: { code: string; name: string; default: boolean }[];
    marketExceptions: MarketExceptions;
};

export enum PaymentLogo {
    Visa = "visa",
    Mastercard = "mastercard",
    Amex = "amex",
    Discover = "discover",
    PayPal = "paypal",
    JCB = "jcb"
}

export type SupportedCountryConfig = {
    isSupported: true;
    productConfig: ProductConfig;
    currency: Currency | null;
    paymentLogos: PaymentLogo[];
    dateFormat: "YMD" | "DMY" | "MDY";
    phonePrefix?: string;
    alwaysShowGuide?: boolean;
    isNoPurchaseMarket?: boolean;
    addressFormat?: string;
} & BaseCountryConfig;

export type UnsupportedCountryConfig = Omit<
    Partial<SupportedCountryConfig>,
    "isSupported"
> & {
    isSupported: false;
    redirect?: string;
} & BaseCountryConfig;

export type CountryConfig = SupportedCountryConfig | UnsupportedCountryConfig;

type OmittedProps =
    | "alpha2"
    | "alpha3"
    | "name"
    | "phonePrefix"
    | "isSupported"; // List the properties to omit
export type TemplateSupportedCountryConfig = {
    [K in keyof SupportedCountryConfig as K extends OmittedProps
        ? never
        : K]: SupportedCountryConfig[K];
};

export type Locale = {
    countryCode: Alpha2 | null;
    language: string | null;
};
