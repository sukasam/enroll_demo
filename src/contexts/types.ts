import { Alpha2 } from "Constants/countryConfig/enums";
import { Product } from "types/Product";
import { OrderResult } from "./types/OrderContextTypes";

export type Settings = {
    locale: string | null;
    version: string;
    cachedCatalogLocale: string | null;
};

type BaseUser = {
    firstName?: string;
    lastName?: string;
    email: string;
    birthDate: string;
    fullName: string;
    phoneNumber: string;
};

export type EnrollmentUser = BaseUser & {
    taxId?: string;
    secondaryTaxId?: string;
    address1: string;
    address2: string;
    city: string;
    state: string;
    zip: string;
    country: Alpha2;
    password: string;
    coapplicantFirstName: string;
    coapplicantLastName: string;
    newsletter: boolean;
    enroller: string | null;
    sponsor: string | null;
    commission: "paylution" | "direct_deposit";
    iban: string;
    bankName: string;
    accountHolder: string;
    bic: string;
};

export type UpgradeUser = BaseUser & {
    id: string;
    href: string;
    email: string;
    shippingAddress: object;
    type: string;
};

export type AppContextType = {
    loading: boolean;
    setLoading: (value: boolean) => void;
    locale: string | null;
    setLocale: (value: string) => void;
    version: string;
    selectedSku: string | null;
    setSelectedSku: (value: string) => void;
    enrollerId: string | null;
    setEnrollerId: (value: string | null) => void;
    sponsorId: string | null;
    setSponsorId: (value: string | null) => void;
    products: Product[];
    setProducts: (value: Product[]) => void;
    cachedCatalogLocale: string | null;
    setCachedCatalogLocale: (value: string) => void;
    userData: EnrollmentUser | UpgradeUser | null;
    setUserData: (value: EnrollmentUser | UpgradeUser) => void;
    userToken: string;
    setUserToken: (token: string) => void;
    orderResult: OrderResult | null;
    setOrderResult: (result: OrderResult) => void;
};
