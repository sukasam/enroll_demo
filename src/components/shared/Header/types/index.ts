export type HeaderProps = {
    hideLocale: boolean;
    referralBar: boolean;
};
