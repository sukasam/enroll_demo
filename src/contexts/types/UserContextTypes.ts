import { AddressType } from "./OrderContextTypes";

export type UserData = {
    firstName?: string;
    lastName?: string;
    email: string;
    fullName: string;
    phoneNumber: string;
    unicityId: string;
};

export type UserContextState = {
    userToken: string | null;
    userData: UserData | null;
    href: string | null;
    enrollerId: string | null;
    sponsorId: string | null;
    enrollerFullName: string | null;
    sponsorFullName: string | null;
    enrollerAllFullName: any | null;
    isLoggedIn: boolean;
    userLocale: string | null;
    isUserContextLoading: boolean;
    activeAccordionSection: number | null;
    mainAddress: AddressType | null;
    setUserToken: (token: string | null) => void;
    setUserData: (data: UserData | null) => void;
    setHref: (href: string | null) => void;
    setEnrollerId: (id: string | null) => void;
    setSponsorId: (id: string | null) => void;
    setEnrollerFullName: (name: string | null) => void;
    setSponsorFullName: (name: string | null) => void;
    setEnrollerAllFullName: (name: any | null) => void;
    setIsLoggedIn: (loggedIn: boolean) => void;
    setUserLocale: (locale: string | null) => void;
    setIsUserContextLoading: (loading: boolean) => void;
    setActiveAccordionSection: (section: number | null) => void;
    goToNextSection: () => void;
    setLoggedInData: (token: string) => void;
    setMainAddress: (address: AddressType | null) => void;
    deleteAuthCookie: () => void;
    resetBacktoDefault: () => void;
};
