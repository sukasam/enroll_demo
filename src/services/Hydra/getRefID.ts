import { Alpha2 } from "Constants/countryConfig/enums";
import httpTools from "Shared/httpTools";

export type Customer = {
    humanName: {
        firstName: string;
        lastName: string;
        fullName: string;
        "fullName@fr": string;
    };
    preferredName?: {
        firstName: string;
        lastName: string;
    };
    mainAddress: {
        country: string;
    };
    unicity: number;
    market: Alpha2;
    href: string;
    customerSite: {
        href: string;
    };
    id: {
        unicity: string;
    };
    profilePicture: {
        href: string;
    };
    businessEntity: {
        companyName: string;
        legalType: string;
    };
    joinDate: string;
    status: string;
    type: string;
    entryPeriod: string;
    referralCode: string;
    subscriptions: {
        type: string;
        endDate: string;
    };
};

export type Response = {
    items?: Customer[];
    ok?: boolean;
};

type GetRefIDParams = {
    country: string;
    language: string;
    refID: string;
};

const getRefID = ({
    country,
    language,
    refID
}: GetRefIDParams): Promise<Response> => {
    const req = {
        method: "GET",
        url: `/customers?referralCode=${refID}&expand=customer,preferredName`,
        headers: {
            "Accept-Language": `${language}-${country}`
        },
        withAuth: false
    };
    return httpTools.sendRequest(req);
};

export default getRefID;
