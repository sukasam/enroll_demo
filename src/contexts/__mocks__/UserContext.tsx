import { AddressType } from "Contexts/types/OrderContextTypes";

export const useUser = jest.fn().mockReturnValue({
    userToken: null,
    setUserToken: jest.fn(),
    enrollerId: null,
    setEnrollerId: jest.fn(),
    sponsorId: null,
    setSponsorId: jest.fn(),
    userData: null,
    setUserData: jest.fn(),
    enrollerFullName: null,
    setEnrollerFullName: jest.fn(),
    sponsorFullName: null,
    setSponsorFullName: jest.fn(),
    isLoggedIn: false,
    setIsLoggedIn: jest.fn(),
    href: null,
    setHref: jest.fn(),
    userLocale: null,
    setUserLocale: jest.fn(),
    isUserContextLoading: false, // almost all test need this to be false vs just the loading test scenario
    setIsUserContextLoading: jest.fn(),
    activeAccordionSection: null,
    setActiveAccordionSection: jest.fn(),
    goToNextSection: jest.fn(),
    setLoggedInData: jest.fn(),
    mainAddress: null as AddressType | null,
    setMainAddress: jest.fn(),
    deleteAuthCookie: jest.fn(),
    resetBacktoDefault: jest.fn()
});

export default useUser;
