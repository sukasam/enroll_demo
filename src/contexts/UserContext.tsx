import { dismissAllErrorToasts } from "Components/shared/ShowToaster";
import { AddressType } from "Contexts/types/OrderContextTypes";
import { UserContextState, UserData } from "Contexts/types/UserContextTypes";
import { useRouter } from "next/router";
import {
    ReactElement,
    ReactNode,
    createContext,
    useCallback,
    useContext,
    useEffect,
    useMemo,
    useState
} from "react";
import useLoginHooks from "./hooks/LoginHooks";

const defaultState: UserContextState = {
    userToken: null,
    setUserToken: () => null,
    enrollerId: null,
    setEnrollerId: () => null,
    sponsorId: null,
    setSponsorId: () => null,
    userData: null,
    setUserData: () => null,
    enrollerFullName: null,
    setEnrollerFullName: () => null,
    sponsorFullName: null,
    setSponsorFullName: () => null,
    enrollerAllFullName: null,
    setEnrollerAllFullName: () => null,
    isLoggedIn: false,
    setIsLoggedIn: () => null,
    href: null,
    setHref: () => null,
    userLocale: null,
    setUserLocale: () => null,
    isUserContextLoading: true,
    setIsUserContextLoading: () => null,
    activeAccordionSection: null,
    setActiveAccordionSection: () => null,
    goToNextSection: () => null,
    setLoggedInData: () => null,
    mainAddress: null as AddressType | null, // Update type
    setMainAddress: () => null,
    deleteAuthCookie: () => null,
    resetBacktoDefault: () => null
};

const UserContext = createContext<UserContextState>(defaultState);

export function UserProvider({
    children,
    testValues
}: {
    children: ReactNode;
    testValues?: Record<string, string>;
}): ReactElement {
    const [userToken, setUserToken] = useState<string | null>(
        defaultState.userToken
    );
    const [enrollerId, setEnrollerId] = useState<string | null>(
        defaultState.enrollerId
    );
    const [sponsorId, setSponsorId] = useState<string | null>(
        defaultState.sponsorId
    );
    const [userData, setUserData] = useState<UserData | null>(
        defaultState.userData
    );
    const [enrollerFullName, setEnrollerFullName] = useState<string | null>(
        defaultState.enrollerFullName
    );
    const [sponsorFullName, setSponsorFullName] = useState<string | null>(
        defaultState.sponsorFullName
    );
    const [enrollerAllFullName, setEnrollerAllFullName] = useState<any | null>(
        defaultState.enrollerAllFullName
    );
    const [isLoggedIn, setIsLoggedIn] = useState<boolean>(
        defaultState.isLoggedIn
    );
    const [href, setHref] = useState<string | null>(defaultState.href);
    const [userLocale, setUserLocale] = useState<string | null>(
        defaultState.userLocale
    );
    const [isUserContextLoading, setIsUserContextLoading] = useState<boolean>(
        defaultState.isUserContextLoading
    );
    const [activeAccordionSection, setActiveAccordionSection] = useState<
        number | null
    >(defaultState.activeAccordionSection);
    const [mainAddress, setMainAddress] = useState<AddressType | null>(
        defaultState.mainAddress
    );

    const router = useRouter();
    const {
        getUserTokenFromCookie,
        setAuthCookie,
        deleteAuthCookie: _deleteAuthCookie,
        setLoggedInData: _setLoggedInData
    } = useLoginHooks();

    const resetBacktoDefault = useCallback(() => {
        setUserToken(defaultState.userToken);
        setEnrollerId(defaultState.enrollerId);
        setSponsorId(defaultState.sponsorId);
        setUserData(defaultState.userData);
        setEnrollerFullName(defaultState.enrollerFullName);
        setSponsorFullName(defaultState.sponsorFullName);
        setEnrollerAllFullName(defaultState.enrollerAllFullName);
        setIsLoggedIn(defaultState.isLoggedIn);
        setHref(defaultState.href);
        setUserLocale(defaultState.userLocale);
        setIsUserContextLoading(defaultState.isUserContextLoading);
        setActiveAccordionSection(defaultState.activeAccordionSection);
        setMainAddress(defaultState.mainAddress);
    }, [
        setUserToken,
        setEnrollerId,
        setSponsorId,
        setUserData,
        setEnrollerFullName,
        setSponsorFullName,
        setEnrollerAllFullName,
        setIsLoggedIn,
        setHref,
        setUserLocale,
        setIsUserContextLoading,
        setActiveAccordionSection,
        setMainAddress
    ]);

    const deleteAuthCookie = useCallback(() => {
        _deleteAuthCookie();
        resetBacktoDefault();
    }, [_deleteAuthCookie, resetBacktoDefault]);

    const setLoggedInData = useCallback(
        async (token: string): Promise<void> => {
            try {
                await _setLoggedInData({
                    userToken: token,
                    setUserData,
                    setHref,
                    setUserToken,
                    setEnrollerFullName,
                    setSponsorFullName,
                    setEnrollerId,
                    setSponsorId,
                    setMainAddress
                });
            } catch (error) {
                deleteAuthCookie();
                resetBacktoDefault();
                throw error;
            }
        },
        [
            _setLoggedInData,
            setUserData,
            setHref,
            setUserToken,
            setEnrollerFullName,
            setSponsorFullName,
            setEnrollerId,
            setSponsorId,
            setMainAddress,
            resetBacktoDefault
        ]
    );

    useEffect(() => {
        setIsLoggedIn(!!userToken);
        setIsUserContextLoading(false);
    }, [userToken]);

    const loadCachedUserData = async (): Promise<void> => {
        setIsUserContextLoading(true);
        const _userToken = getUserTokenFromCookie();
        setUserToken(_userToken);

        const cachedData = sessionStorage.getItem("userContext");
        if (cachedData) {
            try {
                const parsedData: UserContextState = JSON.parse(cachedData);

                setEnrollerId(parsedData.enrollerId);
                setSponsorId(parsedData.sponsorId);
                setUserData(parsedData.userData);
                setEnrollerFullName(parsedData.enrollerFullName);
                setSponsorFullName(parsedData.sponsorFullName);
                setEnrollerAllFullName(parsedData.enrollerAllFullName);
                setIsLoggedIn(!!_userToken);
                setHref(parsedData.href);
                setUserLocale(parsedData.userLocale);
                setMainAddress(parsedData.mainAddress);

                if (!parsedData.href && _userToken) {
                    await setLoggedInData(_userToken);
                }
            } catch (error) {
                console.error(
                    "Failed to parse user context data from sessionStorage:",
                    error
                );
            }
        }
        setIsUserContextLoading(false);
    };

    useEffect(() => {
        if (router.pathname !== "/logout" && router.pathname !== "/reset")
            loadCachedUserData();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const updateCachedUserData = useCallback((): void => {
        setAuthCookie(userToken);
        const dataToCache = {
            enrollerId,
            sponsorId,
            userData,
            enrollerFullName,
            sponsorFullName,
            enrollerAllFullName,
            isLoggedIn,
            href,
            userLocale,
            activeAccordionSection,
            mainAddress
        };
        sessionStorage.setItem("userContext", JSON.stringify(dataToCache));
    }, [
        setAuthCookie,
        userToken,
        enrollerId,
        sponsorId,
        userData,
        enrollerFullName,
        sponsorFullName,
        enrollerAllFullName,
        isLoggedIn,
        href,
        userLocale,
        activeAccordionSection,
        mainAddress
    ]);

    useEffect(() => {
        updateCachedUserData();
    }, [updateCachedUserData]);

    const goToNextSection = useCallback(() => {
        dismissAllErrorToasts();
        setActiveAccordionSection(prevSection =>
            prevSection !== null ? prevSection + 1 : 1
        );
    }, [setActiveAccordionSection]);

    const value = useMemo(
        () => ({
            userToken,
            setUserToken,
            enrollerId,
            setEnrollerId,
            sponsorId,
            setSponsorId,
            userData,
            setUserData,
            enrollerFullName,
            setEnrollerFullName,
            sponsorFullName,
            setSponsorFullName,
            enrollerAllFullName,
            setEnrollerAllFullName,
            isLoggedIn,
            setIsLoggedIn,
            href,
            setHref,
            userLocale,
            setUserLocale,
            isUserContextLoading,
            setIsUserContextLoading,
            activeAccordionSection,
            setActiveAccordionSection,
            goToNextSection,
            setLoggedInData,
            mainAddress,
            setMainAddress,
            deleteAuthCookie,
            resetBacktoDefault
        }),
        [
            userToken,
            enrollerId,
            sponsorId,
            userData,
            enrollerFullName,
            sponsorFullName,
            enrollerAllFullName,
            isLoggedIn,
            href,
            userLocale,
            isUserContextLoading,
            activeAccordionSection,
            goToNextSection,
            setLoggedInData,
            mainAddress,
            setMainAddress,
            deleteAuthCookie,
            resetBacktoDefault
        ]
    );

    const providerValue = useMemo(
        () => ({
            ...value,
            ...testValues
        }),
        [value, testValues]
    );

    return (
        <UserContext.Provider value={providerValue}>
            {children}
        </UserContext.Provider>
    );
}

export const useUser = (): UserContextState => useContext(UserContext);
