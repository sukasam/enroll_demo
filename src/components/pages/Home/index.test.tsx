import { render, RenderResult, screen, waitFor } from "@testing-library/react";
import { UserContextState } from "Contexts/types/UserContextTypes";
import { UserProvider, useUser } from "Contexts/UserContext";
import { getCookie, setCookie } from "cookies-next";
import { Customer } from "Hydra/getRefID";
import { GetServerSidePropsContext } from "next";
import mixpanelService from "Services/mixpanel/initializeMixPanel";
import Landing, { getServerSideProps, HomeServerProps } from "./index";

// Mocks
jest.mock("next/router", () => ({
    useRouter: jest.fn(() => ({ prefetch: jest.fn().mockResolvedValue(true) }))
}));

jest.mock("next/image", () => ({
    __esModule: true,
    default: ({
        src,
        alt,
        "data-testid": testId,
        ...props
    }: {
        src: string;
        alt: string;
        "data-testid"?: string;
        [key: string]: any;
    }): JSX.Element => (
        // eslint-disable-next-line @next/next/no-img-element
        <img
            src={src}
            alt={alt}
            data-testid={testId || "next-image"}
            {...props}
        />
    )
}));

jest.mock("Services/mixpanel/initializeMixPanel", () => ({
    trackPageView: jest.fn(),
    registerSuperProperties: jest.fn()
}));

jest.mock("Contexts/UserContext", () => ({
    ...jest.requireActual("Contexts/UserContext"),
    useUser: jest.fn()
}));

jest.mock("cookies-next", () => ({
    getCookie: jest.fn(),
    setCookie: jest.fn(),
    deleteCookie: jest.fn()
}));

jest.mock("Services/geoip", () => ({
    __esModule: true,
    default: jest.fn().mockResolvedValue({ alpha2: "US", language: "en" })
}));

jest.mock("Services/utils/refValidation", () => ({
    __esModule: true,
    default: jest.fn().mockResolvedValue({
        isValid: true,
        unicityID: "12345",
        refResults: { humanName: { firstName: "John", lastName: "Doe" } }
    })
}));

jest.mock("./components/HomeContent", () => ({
    __esModule: true,
    default: jest.fn(() => <div data-testid="mock-home-content" />)
}));

const mockUseUser = useUser as jest.MockedFunction<typeof useUser>;

describe("Landing", () => {
    const defaultProps: HomeServerProps = {
        referrer: null,
        refId: null,
        country: "US",
        language: "en"
    };
    const setup = (props: Partial<HomeServerProps> = {}): RenderResult =>
        render(
            <UserProvider>
                <Landing {...{ ...defaultProps, ...props }} />
            </UserProvider>
        );

    beforeEach(() => {
        jest.clearAllMocks();
        mockUseUser.mockReturnValue({
            setEnrollerId: jest.fn(),
            setSponsorId: jest.fn(),
            setEnrollerFullName: jest.fn(),
            setSponsorFullName: jest.fn(),
            isUserContextLoading: false
        } as Partial<UserContextState> as UserContextState);
    });

    it("renders correctly without referrer", () => {
        const { asFragment } = setup();
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly with referrer", () => {
        const referrer: Partial<Customer> = {
            humanName: {
                firstName: "John",
                lastName: "Doe",
                fullName: "John Doe",
                "fullName@fr": "John Doe"
            },
            mainAddress: { country: "US" },
            unicity: 12345,
            href: "http://example.com/customer"
        };
        const { asFragment } = setup({
            referrer: referrer as Customer,
            refId: "123"
        });
        expect(asFragment()).toMatchSnapshot();
    });

    it("tracks page view", async () => {
        setup();
        await waitFor(() => {
            expect(mixpanelService.trackPageView).toHaveBeenCalledWith("home");
        });
    });

    it("renders HomeContent and Aside components", () => {
        setup();
        expect(screen.getByTestId("mock-home-content")).toBeInTheDocument();
        expect(
            screen.getByTestId("starting_business_just_become_easier_section")
        ).toBeInTheDocument();
    });

    it("sets user context when referrer and refId are provided", () => {
        const referrer: Partial<Customer> = {
            humanName: {
                firstName: "John",
                lastName: "Doe",
                fullName: "John Doe",
                "fullName@fr": "John Doe"
            },
            unicity: 12345
        };
        setup({ referrer: referrer as Customer, refId: "12345" });
        expect(mockUseUser().setEnrollerId).toHaveBeenCalledWith("12345");
        expect(mockUseUser().setSponsorId).toHaveBeenCalledWith("12345");
        expect(mockUseUser().setEnrollerFullName).toHaveBeenCalledWith(
            "John Doe"
        );
        expect(mockUseUser().setSponsorFullName).toHaveBeenCalledWith(
            "John Doe"
        );
    });
});

describe("getServerSideProps", () => {
    const mockContext = { req: {} } as GetServerSidePropsContext;

    beforeEach(() => {
        jest.clearAllMocks();
        (getCookie as jest.Mock).mockImplementation(
            (key: string) =>
                ({ country: "US", language: "en", refId: null }[key] || null)
        );
    });

    it("returns null referrer and refId when no refId cookie", async () => {
        const result = await getServerSideProps(mockContext);
        expect(result).toEqual({
            props: {
                country: "US",
                language: "en",
                referrer: null,
                refId: null
            }
        });
    });

    it("sets country and language cookies based on geolocation", async () => {
        await getServerSideProps(mockContext);
        expect(setCookie).toHaveBeenCalledWith("country", "US", mockContext);
        expect(setCookie).toHaveBeenCalledWith("language", "en", mockContext);
    });
});
