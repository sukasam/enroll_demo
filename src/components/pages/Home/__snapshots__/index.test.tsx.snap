// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing renders correctly with referrer 1`] = `
<DocumentFragment>
  <div
    class=MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-XXXXX-MuiGrid-root
  >
    <div
      class=MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8 css-XXXXX-MuiGrid-root
    >
      <div
        data-testid="mock-home-content"
      />
    </div>
    <div
      class=MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 aside-wrapper css-XXXXX-MuiGrid-root
      data-testid="starting_business_just_become_easier_section"
    >
      <div
        class="tile aside"
      >
        <div
          class="banner"
        >
          <div
            class="overlay"
            data-testid="home_page_overlay"
          />
          <div
            class="content"
          >
            <div
              class="title"
              data-testid="right_column_starting_your_business_just_got_easier"
            >
              right_column_starting_your_business_just_got_easier
            </div>
          </div>
        </div>
        <div
          class="title"
          data-testid="right_column_your_business_your_way"
        >
          right_column_your_business_your_way
        </div>
        <p
          data-testid="right_column_unicity_help_become_leader"
        >
          right_column_unicity_help_become_leader
        </p>
        <ul>
          <li
            data-testid="right_column_wholesale_product_pricing"
          >
            right_column_wholesale_product_pricing
          </li>
          <li
            data-testid="right_column_earn_commissions_product_sales"
          >
            right_column_earn_commissions_product_sales
          </li>
          <li
            data-testid="right_column_build_your_business_your_way"
          >
            right_column_build_your_business_your_way
          </li>
          <li
            data-testid="right_column_growth_incentives"
          >
            right_column_growth_incentives
          </li>
          <li
            data-testid="right_column_training_education"
          >
            right_column_training_education
          </li>
          <li
            data-testid="right_column_business_building_tools"
          >
            right_column_business_building_tools
          </li>
          <li
            data-testid="right_column_networking_opportunities"
          >
            right_column_networking_opportunities
          </li>
          <li
            data-testid="right_column_be_part_of_community"
          >
            right_column_be_part_of_community
          </li>
        </ul>
        <a
          data-testid="right_column_learn_more"
          href="right_column_learn_more_link"
          rel="noreferrer noopener"
          target="_blank"
        >
          right_column_learn_more
          <img
            alt=""
            data-testid="next-image"
            height="16"
            quality="85"
            src="/svg/external-link.svg"
            width="16"
          />
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Landing renders correctly without referrer 1`] = `
<DocumentFragment>
  <div
    class=MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-XXXXX-MuiGrid-root
  >
    <div
      class=MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8 css-XXXXX-MuiGrid-root
    >
      <div
        data-testid="mock-home-content"
      />
    </div>
    <div
      class=MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 aside-wrapper css-XXXXX-MuiGrid-root
      data-testid="starting_business_just_become_easier_section"
    >
      <div
        class="tile aside"
      >
        <div
          class="banner"
        >
          <div
            class="overlay"
            data-testid="home_page_overlay"
          />
          <div
            class="content"
          >
            <div
              class="title"
              data-testid="right_column_starting_your_business_just_got_easier"
            >
              right_column_starting_your_business_just_got_easier
            </div>
          </div>
        </div>
        <div
          class="title"
          data-testid="right_column_your_business_your_way"
        >
          right_column_your_business_your_way
        </div>
        <p
          data-testid="right_column_unicity_help_become_leader"
        >
          right_column_unicity_help_become_leader
        </p>
        <ul>
          <li
            data-testid="right_column_wholesale_product_pricing"
          >
            right_column_wholesale_product_pricing
          </li>
          <li
            data-testid="right_column_earn_commissions_product_sales"
          >
            right_column_earn_commissions_product_sales
          </li>
          <li
            data-testid="right_column_build_your_business_your_way"
          >
            right_column_build_your_business_your_way
          </li>
          <li
            data-testid="right_column_growth_incentives"
          >
            right_column_growth_incentives
          </li>
          <li
            data-testid="right_column_training_education"
          >
            right_column_training_education
          </li>
          <li
            data-testid="right_column_business_building_tools"
          >
            right_column_business_building_tools
          </li>
          <li
            data-testid="right_column_networking_opportunities"
          >
            right_column_networking_opportunities
          </li>
          <li
            data-testid="right_column_be_part_of_community"
          >
            right_column_be_part_of_community
          </li>
        </ul>
        <a
          data-testid="right_column_learn_more"
          href="right_column_learn_more_link"
          rel="noreferrer noopener"
          target="_blank"
        >
          right_column_learn_more
          <img
            alt=""
            data-testid="next-image"
            height="16"
            quality="85"
            src="/svg/external-link.svg"
            width="16"
          />
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
